### PR TITLE
Initial physics file reader for CAM-SIMA

### PR DESCRIPTION
**Pull request by nusbaume**
Tag name (required for release branches):
Originator(s): @mwaxmonsky @nusbaume

Description (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):

This PR contains the CAM-SIMA subroutines and interfaces required for implementing the file I/O reader object in PR ESCOMP/atmospheric_physics#240 in a CAM-SIMA simulation.

Currently CAM-SIMA uses the Parallel I/O (PIO) system to read NetCDF files whenever a user calls the FileIO object, and so that is what is used for this initial implementation, but if different methods are required in the future we can hopefully implement them in a similar but independent fashion.

This PR also adds a new MUSICA namelist variable to fix a bug that prevented the model from building whenever the MUSICA CCPP scheme wasn't being used.

Describe any changes made to build system: N/A

Describe any changes made to the namelist: N/A

List any changes to the defaults for the input datasets (e.g. boundary datasets): N/A

List all files eliminated and why: N/A

List all files added and what they do:

A       src/physics/utils/pio_reader.F90
A       src/physics/utils/pio_reader_sub.F90
 - SIMA's NetCDF file reader object implementation using PIO.

A       src/physics/utils/rrtmgp-data
 - RRTMGP data files.  Mostly used for testing now, but will be needed for RRTMGP itself once the time comes (otherwise it can simply be removed).

A       src/physics/utils/musica_sima_namelist.F90
 - Implement new CAM-SIMA 'musica_config' namelist variable in a separate, non-CCPPized module.

List all existing files that have been modified, and describe the changes: 
(Helpful git command: `git diff --name-status development...<your_branch_name>`)

M       .gitmodules
M       src/physics/ncar_ccpp
 - Updated atmospheric_physics submodule

M       cime_config/cam_autogen.py
 - Make sure the 'phys_utils' directory is always available during a build.

M       cime_config/namelist_definition_cam.xml
M       src/control/runtime_opts.F90
M       src/physics/utils/musica_ccpp_dependencies.F90
 - Implement the 'musica_config' SIMA namelist option to control MUSICA dependencies settings.

M       src/utils/string_utils.F90
 - Make 'to_lower' and 'to_upper' functions public.

If there are new failures (compared to the `test/existing-test-failures.txt` file),
have them OK'd by the gatekeeper, note them here, and add them to the file.
If there are baseline differences, include the test and the reason for the
diff. What is the nature of the change? Roundoff?

derecho/intel/aux_sima:

derecho/gnu/aux_sima:

If this changes climate describe any run(s) done to evaluate the new
climate in enough detail that it(they) could be reproduced:

CAM-SIMA date used for the baseline comparison tests if different than latest:

### CODE REVIEW

**Review by K20shores — APPROVED** [review]

File: src/physics/utils/pio_reader.F90 [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+
+    !Error code parameters
+    integer, parameter :: pio_inq_dim_id_err      = 1
+    integer, parameter :: pio_inq_dim_len_err     = 2
+    integer, parameter :: pio_inq_var_id_err      = 3
+    integer, parameter :: pio_inq_var_info_err    = 4
+    integer, parameter :: bad_var_rank_err        = 5
+    integer, parameter :: too_high_rank_err       = 6
+    integer, parameter :: pio_get_var_err         = 7
+    integer, parameter :: not_char_type_err       = 8
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90:19]: Hi @jimmielin, in general the only requirement I have seen with regards to error codes in the CCPP is that a non-zero error code means an error occurred, and a zero error code means no error.  However one could imagine a benefit if there were universally-designated error codes in the CCPP (then a physics scheme could alter it's behavior depending on what the error code was, instead of just ending the model simulation as is done now).

However, getting "universal" error codes would likely require some discussions with the various CCPP stakeholders before anything was designed or implemented, so it wasn't something I wanted to fully tackle here.  Instead I just wanted the error codes to be internally consistent within the CAM-SIMA `pio_reader` schemes.  Although this may not currently benefit physics, it could benefit CAM-SIMA itself if we start using these routines throughout the host model code instead of direct calls to `pio_get_var` (which would help centralize CAM-SIMA I/O routines).

Hopefully that helps explain the reasoning here?  Of course if you feel differently, or have any other questions or concerns, then just let me know!

File: src/physics/utils/pio_reader.F90 [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+
+    type :: sima_pio_fh_type
+        logical            :: is_file_open = .false.  !Is NetCDF file currently open?
+        type(file_desc_t)  :: pio_fh                  !PIO File handle type
+        character(len=cl)  :: file_path              !Local path to NetCDF file
+    end type
```
Comment (mattldawson) [inline src/physics/utils/pio_reader.F90:25]: Since this is a private class in this module, would a name like `file_handle_t` be better? I took me a minute to figure out what `fh` was, and I think the `sima` and `pio` will be pretty obvious from the context.

File: src/physics/utils/pio_reader.F90 [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+
+    subroutine close_netcdf_file(this, errcode, errmsg)
+        use pio, only: pio_closefile
+
+        class(pio_reader_t), intent(inout)  :: this
+        integer,             intent(out) :: errcode
+        character(len=*),    intent(out) :: errmsg
+
+        !Close NetCDF File:
+        call pio_closefile(this%sima_pio_fh%pio_fh)
```
Comment (mattldawson) [inline src/physics/utils/pio_reader.F90:82]: Should you first check if `is_file_open == .true.`?

File: src/physics/utils/pio_reader.F90 [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+        integer              :: err_handling    !PIO error handling code
+        integer              :: var_id          !NetCDF variable ID
+        integer              :: ndims           !Number of variable dimensions on NetCDF file
+        integer, allocatable :: dim_ids(:)      !Variable dimension IDs
+        integer, allocatable :: dim_sizes(:)    !Variable dimension sizes
+  
+        integer :: i !loop control variable
+        !----------------------
+  
+        !Extract open file information:
```
Comment (mattldawson) [inline src/physics/utils/pio_reader.F90:121]: Should there be a check that `is_file_open == .true.`? (similar comment for remaining functions)

**Review by mattldawson — APPROVED** [review]
Looks great! Just a couple minor comments.

The only other this is that I would suggest adding some tests for this implementation. Seems like there could be quite a few corner cases given all the conditional blocks.

File: src/physics/utils/pio_reader.F90 [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+        character(len=*),    intent(out) :: errmsg
+        
+        character(len=cl)  :: local_file_path  !NetCDF file path on local file system
+
+        if(this%sima_pio_fh%is_file_open) then
+          errcode = 1
+          errmsg = "Trying to reuse pio_reader already used for: '"//this%sima_pio_fh%file_path
+        end if
+
+        if(file_path .eq. 'UNSET_PATH') then
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90:57]: Yes, `UNSET_PATH` is a SIMA-specific keyword that is the default value whenever a namelist-specified file path is not set, as defined in CAM-SIMA here:

https://github.com/ESCOMP/CAM-SIMA/blob/development/cime_config/namelist_definition_cam.xml#L190

You raise a good point though about this being problematic if brought back into CAM.  Maybe we can just remove it here with the assumption that it is up to the caller to ensure that the path is "real"?  Otherwise the routines here would just exit whenever PIO tried to do something with the file.  Does that sound reasonable to you?

File: src/physics/utils/pio_reader.F90 (hunk 2 above)
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90]: Works for me, done!

File: src/physics/utils/pio_reader.F90 (hunk 3 above)
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90:91]: Good call!  Done!

File: src/physics/utils/pio_reader.F90 (hunk 4 above)
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90:138]: Yes, good catch!  Added checks to all read subroutines.

**Review by boulderdaze — APPROVED** [review]
Looks good!

File: src/physics/utils/pio_reader_sub.F90 [hunk 6]
```diff
@@ -0,0 +1,8 @@
+submodule (ccpp_io_reader) pio_submodule
+use pio_reader, only: pio_reader_t
+implicit none
+contains
+  module procedure create_netcdf_reader_t
+    allocate(pio_reader_t :: r)
+  end procedure create_netcdf_reader_t
+end submodule
```
Comment (kuanchihwang) [inline src/physics/utils/pio_reader_sub.F90:8]:
```suggestion
end submodule pio_submodule
```

Include the submodule name in the `end submodule` statement.

File: src/physics/utils/pio_reader.F90 [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+
+    type :: file_handle_t
+        logical            :: is_file_open = .false.  !Is NetCDF file currently open?
+        type(file_desc_t)  :: pio_fh                  !PIO File handle type
+        character(len=cl)  :: file_path              !Local path to NetCDF file
+    end type
+
+    type, extends(abstract_netcdf_reader_t) :: pio_reader_t
+    private
+        type(file_handle_t) :: sima_pio_fh !PIO File handle type
```
Comment (kuanchihwang) [inline src/physics/utils/pio_reader.F90:30]:
```suggestion
        private
        type(file_handle_t) :: sima_pio_fh !PIO File handle type
```

Increase indent.

**Review by kuanchihwang — COMMENTED** [review]
I have a few comments/suggestions for your consideration. Thank you!

File: src/physics/utils/pio_reader_sub.F90 (hunk 6 above)
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader_sub.F90]: Added!

File: src/physics/utils/pio_reader.F90 [hunk 8]
```diff
@@ -0,0 +1,751 @@
...
+    integer, parameter :: pio_inq_var_info_err    = 4
+    integer, parameter :: bad_var_rank_err        = 5
+    integer, parameter :: too_high_rank_err       = 6
+    integer, parameter :: pio_get_var_err         = 7
+    integer, parameter :: not_char_type_err       = 8
+
+    type :: sima_pio_fh_type
+        logical            :: is_file_open = .false.  !Is NetCDF file currently open?
+        type(file_desc_t)  :: pio_fh                  !PIO File handle type
+        character(len=cl)  :: file_path              !Local path to NetCDF file
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90]: Good catch, fixed!

File: src/physics/utils/pio_reader.F90 [hunk 9]
```diff
@@ -0,0 +1,751 @@
...
+        character(len=cl)  :: local_file_path  !NetCDF file path on local file system
+
+        if(this%sima_pio_fh%is_file_open) then
+          errcode = 1
+          errmsg = "Trying to reuse pio_reader already used for: '"//this%sima_pio_fh%file_path
+        end if
+
+        if(file_path .eq. 'UNSET_PATH') then
+         errcode = 1
+         errmsg = "Found UNSET_PATH trying to open file"
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90:61]: Added!

File: src/physics/utils/pio_reader.F90 [hunk 10]
```diff
@@ -0,0 +1,751 @@
...
+        class(pio_reader_t), intent(inout)  :: this
+        character(len=*),    intent(in)  :: file_path
+        integer,             intent(out) :: errcode
+        character(len=*),    intent(out) :: errmsg
+        
+        character(len=cl)  :: local_file_path  !NetCDF file path on local file system
+
+        if(this%sima_pio_fh%is_file_open) then
+          errcode = 1
+          errmsg = "Trying to reuse pio_reader already used for: '"//this%sima_pio_fh%file_path
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90:55]: Added!

File: src/physics/utils/pio_reader.F90 (hunk 5 above)
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90]: Updated!

File: src/physics/utils/pio_reader.F90 [hunk 11]
```diff
@@ -0,0 +1,751 @@
...
+           errcode = pio_inq_var_info_err !Make sure error code is non-zero
+           errmsg  = "Failed to find number of dimensions for '"//varname//"' in "//file_path
+           !Reset PIO back to original error handling method:
+           call pio_seterrorhandling(pio_file_handle, err_handling)
+           return
+        end if
+  
+        !Check that variable is a character array
+        !(as we cannot currently handle string-type variables):
+        if(nc_type /= NF90_CHAR) then
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90]: Updated!  Sounds good to me; @nusbaume just let me know if you feel differently.

File: src/physics/utils/pio_reader.F90 [hunk 12]
```diff
@@ -0,0 +1,751 @@
...
+        call pio_seterrorhandling(pio_file_handle, err_handling)
+  
+        !Variable was successfully read, so properly set the error
+        !code and message:
+        errcode = 0
+        errmsg = ''
+    end subroutine get_netcdf_var_real
+
+    subroutine get_netcdf_var_char(this, varname, var, errcode, errmsg)
+        use ccpp_kinds, only: kind_phys
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90]: Removed!

File: src/physics/utils/pio_reader.F90 [hunk 13]
```diff
@@ -0,0 +1,751 @@
...
+        this%sima_pio_fh%is_file_open = .false.
+        this%sima_pio_fh%file_path = ''
+
+        !File was successfully closed
+        errcode = 0
+        errmsg = ''
+    end subroutine close_netcdf_file
+
+    subroutine get_netcdf_var_int(this, varname, var, errcode, errmsg)
+        use ccpp_kinds, only: kind_phys
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/physics/utils/pio_reader.F90]: Removed!

File: src/physics/utils/pio_reader.F90 [hunk 14]
```diff
@@ -0,0 +1,751 @@
...
+    implicit none
+    private
+  
+    public :: pio_reader_t
+
+    !Error code parameters
+    integer, parameter :: pio_inq_dim_id_err      = 1
+    integer, parameter :: pio_inq_dim_len_err     = 2
+    integer, parameter :: pio_inq_var_id_err      = 3
+    integer, parameter :: pio_inq_var_info_err    = 4
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90:15]: Thanks for the good suggestion @kuanchihwang!  I've gone ahead and added a new subroutine where the `errmsg` string is set via the `pio_strerror` function, which hopefully will provide more information than the `errmsg` strings I had manually been setting before.

In terms of the error codes themselves, as mentioned above we might have CCPP-specific error codes for various errors which will need to be independent from the underlying I/O implementation (e.g. PIO).  Thus I assumed that defining new error codes at the `pio_reader` level will make it much easier to replace with "official" CCPP error codes whenever we get to that point.

Of course if you disagree with this plan then please let me know!

File: src/physics/utils/pio_reader.F90 (hunk 7 above)
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90]: Fixed!  I also found some other inconsistent indenting, so went through the entire file to make it internally consistent.

**Review by kuanchihwang — APPROVED** [review]
Looks great to me!

**Review by boulderdaze — APPROVED** [review]
